File: tools/tf-types-ts/src/core/bridge-did.ts
/**
 * DID (Decentralized Identifier — W3C DID Core 1.0) bridge.
 *
 * Resolves a DID URL into a DID document, picks a verification method,
 * and projects it into a TrustForge ActorIdentity. Supports the
 * mandatory `did:key` (RFC 9468 / multibase ed25519) method out of the
 * box; arbitrary methods can be plugged in via the resolver callback
 * so callers can wire `did:web`, `did:plc`, `did:ion`, etc. without
 * pulling extra deps.
 */

import { sha256 } from "@noble/hashes/sha256";

import type { ActorIdentity } from "../generated/actor-identity.js";
import { BridgeFailure, type Bridge, type BridgeKind } from "./bridges.js";
import { ed25519Verify } from "./crypto.js";

export interface DidVerificationMethod {
  id: string;
  /** e.g. `Ed25519VerificationKey2020`, `JsonWebKey2020`. */
  type: string;
  controller: string;
  publicKeyMultibase?: string;
  publicKeyJwk?: Record<string, unknown>;
}

export interface DidDocument {
  "@context"?: string | string[];
  id: string;
  controller?: string | string[];
  verificationMethod?: DidVerificationMethod[];
  authentication?: Array<string | DidVerificationMethod>;
  assertionMethod?: Array<string | DidVerificationMethod>;
  service?: Array<{ id: string; type: string; serviceEndpoint: string }>;
}

export type DidResolver = (didUrl: string) => Promise<DidDocument>;

export interface DidBridgeConfig {
  bridgeId: string;
  trustDomain: string;
  /** Resolver for arbitrary DID methods. Defaults to a built-in
   *  `did:key` resolver. Wire in `did:web` / `did:plc` / etc. by
   *  composing your own resolver. */
  resolver?: DidResolver;
  /** Optional allow-list of method names — if set, the bridge refuses
   *  to resolve any method not in this set. */
  allowedMethods?: string[];
}

export interface DidVerificationResult {
  identity: ActorIdentity;
  document: DidDocument;
  /** Verification method id used to populate the actor identity. */
  verificationMethodId: string;
}

export class DidBridge implements Bridge {
  readonly kind: BridgeKind = "did";
  readonly bridgeId: string;
  readonly trustDomain: string;
  private readonly cfg: DidBridgeConfig;

  constructor(cfg: DidBridgeConfig) {
    this.bridgeId = cfg.bridgeId;
    this.trustDomain = cfg.trustDomain;
    this.cfg = cfg;
  }

  /** Resolve a DID URL to a DID document. Tries the built-in did:key
   *  resolver first; falls back to the caller-supplied resolver if any. */
  async resolve(didUrl: string): Promise<DidDocument> {
    const method = parseDidMethod(didUrl);
    if (this.cfg.allowedMethods && !this.cfg.allowedMethods.includes(method)) {
      throw new BridgeFailure({
        code: "rejected",
        message: `DID method ${method} not in allow-list`,
      });
    }
    if (method === "key") {
      return resolveDidKey(didUrl);
    }
    if (this.cfg.resolver) {
      return this.cfg.resolver(didUrl);
    }
    throw new BridgeFailure({
      code: "unsupported",
      message: `no resolver configured for DID method ${method}`,
    });
  }

  /** Resolve a DID URL and project it into an ActorIdentity. Picks the
   *  first authentication method that maps to a known signature
   *  algorithm (ed25519 today; ECDSA / RSA via the JWK projector when
   *  the method declares them). */
  async accept(didUrl: string): Promise<DidVerificationResult> {
    const document = await this.resolve(didUrl);
    const vms = document.verificationMethod ?? [];
    const auth = (document.authentication ?? []).map((a) =>
      typeof a === "string" ? vms.find((v) => v.id === a) : a,
    );
    const candidates = (auth.filter(Boolean) as DidVerificationMethod[]).concat(vms);
    if (candidates.length === 0) {
      throw new BridgeFailure({
        code: "rejected",
        message: "DID document has no verification methods",
      });
    }
    const vm = candidates[0]!;
    const publicKey = extractPublicKey(vm);
    if (!publicKey) {
      throw new BridgeFailure({
        code: "unsupported",
        message: `verification method ${vm.id} has no usable public key`,
      });
    }
    const subject = encodeURIComponent(document.id);
    const identity: ActorIdentity = {
      identity_version: "1",
      actor_id: `tf:actor:human:${this.cfg.trustDomain}/${subject}`,
      actor_type: "human",
      public_keys: [
        {
          key_id: vm.id,
          algorithm: publicKey.algorithm,
          public_key: Buffer.from(publicKey.bytes).toString("base64"),
          purpose: "signing",
        },
      ],
      trust_levels: ["T2"],
      authority_roots: [
        {
          kind: "federation",
          id: vm.controller,
        },
      ],
      valid_from: new Date().toISOString(),
    };
    return { identity, document, verificationMethodId: vm.id };
  }

  /** Verify an ed25519 signature using the resolved DID's first
   *  ed25519 verification method. */
  async verifySignature(
    didUrl: string,
    message: Uint8Array,
    signature: Uint8Array,
  ): Promise<boolean> {
    const doc = await this.resolve(didUrl);
    const vms = doc.verificationMethod ?? [];
    for (const vm of vms) {
      const pk = extractPublicKey(vm);
      if (pk?.algorithm === "ed25519") {
        const ok = await ed25519Verify(pk.bytes, message, signature);
        if (ok) return true;
      }
    }
    return false;
  }
}

function parseDidMethod(did: string): string {
  const m = /^did:([a-z0-9]+):/.exec(did);
  if (!m) throw new BridgeFailure({ code: "invalid-input", message: `not a DID: ${did}` });
  return m[1]!;
}

function extractPublicKey(vm: DidVerificationMethod): { algorithm: string; bytes: Uint8Array } | null {
  if (vm.publicKeyMultibase) {
    const decoded = decodeMultibase(vm.publicKeyMultibase);
    if (!decoded) return null;
    // Per RFC 9468, ed25519 multikeys are prefixed with 0xed 0x01.
    if (decoded[0] === 0xed && decoded[1] === 0x01) {
      return { algorithm: "ed25519", bytes: decoded.slice(2) };
    }
    // Generic fallback: pass the raw bytes through.
    return { algorithm: vm.type.toLowerCase(), bytes: decoded };
  }
  if (vm.publicKeyJwk) {
    const jwk = vm.publicKeyJwk;
    if (jwk.kty === "OKP" && jwk.crv === "Ed25519" && typeof jwk.x === "string") {
      return { algorithm: "ed25519", bytes: base64UrlToBytes(jwk.x) };
    }
    if (jwk.kty === "EC" && typeof jwk.x === "string" && typeof jwk.y === "string") {
      const x = base64UrlToBytes(jwk.x);
      const y = base64UrlToBytes(jwk.y);
      const buf = new Uint8Array(1 + x.length + y.length);
      buf[0] = 0x04;
      buf.set(x, 1);
      buf.set(y, 1 + x.length);
      return { algorithm: jwk.crv === "P-256" ? "p256" : "ec", bytes: buf };
    }
  }
  return null;
}

/** Built-in did:key resolver. Supports the ed25519 multikey prefix
 *  (0xed 0x01) per the W3C did:key specification. */
async function resolveDidKey(didUrl: string): Promise<DidDocument> {
  const m = /^did:key:([a-zA-Z1-9]+)/.exec(didUrl);
  if (!m) throw new BridgeFailure({ code: "invalid-input", message: `not a did:key: ${didUrl}` });
  const multibase = m[1]!;
  const id = `did:key:${multibase}`;
  return {
    "@context": ["https://www.w3.org/ns/did/v1"],
    id,
    verificationMethod: [
      {
        id: `${id}#${multibase}`,
        type: "Ed25519VerificationKey2020",
        controller: id,
        publicKeyMultibase: multibase,
      },
    ],
    authentication: [`${id}#${multibase}`],
    assertionMethod: [`${id}#${multibase}`],
  };
}

/** Decode a multibase string. Supports `z` (base58btc), `m` (base64),
 *  `u` (base64url-no-pad). */
function decodeMultibase(s: string): Uint8Array | null {
  if (s.length === 0) return null;
  const prefix = s[0]!;
  const body = s.slice(1);
  switch (prefix) {
    case "z":
      return base58btcDecode(body);
    case "m":
      return new Uint8Array(Buffer.from(body, "base64"));
    case "u":
      return base64UrlToBytes(body);
    default:
      return null;
  }
}

const BASE58_ALPHABET =
  "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

function base58btcDecode(s: string): Uint8Array | null {
  if (s.length === 0) return new Uint8Array();
  // Count leading '1' chars — each is a leading zero byte.
  let zeros = 0;
  while (zeros < s.length && s[zeros] === "1") zeros += 1;
  // Allocate enough capacity (log(58)/log(256) ≈ 0.733).
  const size = Math.ceil((s.length - zeros) * 0.733) + 1;
  const b256 = new Uint8Array(size);
  for (let i = zeros; i < s.length; i++) {
    const idx = BASE58_ALPHABET.indexOf(s[i]!);
    if (idx < 0) return null;
    let carry = idx;
    for (let j = size - 1; j >= 0; j--) {
      carry += b256[j]! * 58;
      b256[j] = carry & 0xff;
      carry >>= 8;
    }
    if (carry !== 0) return null;
  }
  // Skip leading zeros in b256.
  let start = 0;
  while (start < size && b256[start] === 0) start += 1;
  const out = new Uint8Array(zeros + (size - start));
  out.set(b256.subarray(start), zeros);
  return out;
}

function base64UrlToBytes(b64u: string): Uint8Array {
  let s = b64u.replace(/-/g, "+").replace(/_/g, "/");
  while (s.length % 4 !== 0) s += "=";
  return new Uint8Array(Buffer.from(s, "base64"));
}

/** Compute the ed25519 multibase identifier for a 32-byte ed25519
 *  public key — useful for tests and for callers that need to mint a
 *  did:key for a freshly-generated key. */
export function ed25519PublicKeyToDidKey(pub: Uint8Array): string {
  if (pub.length !== 32) throw new Error("ed25519 public key must be 32 bytes");
  const prefixed = new Uint8Array(2 + 32);
  prefixed[0] = 0xed;
  prefixed[1] = 0x01;
  prefixed.set(pub, 2);
  return `z${base58btcEncode(prefixed)}`;
}

function base58btcEncode(bytes: Uint8Array): string {
  if (bytes.length === 0) return "";
  let zeros = 0;
  while (zeros < bytes.length && bytes[zeros] === 0) zeros += 1;
  const size = Math.ceil(bytes.length * 1.366) + 1;
  const b58 = new Uint8Array(size);
  for (let i = zeros; i < bytes.length; i++) {
    let carry = bytes[i]!;
    for (let j = size - 1; j >= 0; j--) {
      carry += b58[j]! * 256;
      b58[j] = carry % 58;
      carry = Math.floor(carry / 58);
    }
  }
  let start = 0;
  while (start < size && b58[start] === 0) start += 1;
  let out = "";
  for (let i = 0; i < zeros; i++) out += "1";
  for (let i = start; i < size; i++) out += BASE58_ALPHABET[b58[i]!];
  return out;
}

// Touch sha256 so the import doesn't end up dead when the spec adds a
// digest-based identity check in a future version.
void sha256;

File: crates/tf-types/src/bridge_did.rs
//! DID (W3C DID Core 1.0) bridge — Rust mirror of TS.

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::bridges::{Bridge, BridgeError, BridgeKind};
use crate::generated::{
    ActorIdentity, ActorIdentity_IdentityVersion, ActorType, AuthorityRoot, AuthorityRoot_Kind,
    PublicKey, PublicKey_Purpose, TrustLevel,
};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DidVerificationMethod {
    pub id: String,
    #[serde(rename = "type")]
    pub kind: String,
    pub controller: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub public_key_multibase: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub public_key_jwk: Option<Value>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DidDocument {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification_method: Option<Vec<DidVerificationMethod>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub authentication: Option<Vec<Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub controller: Option<Value>,
}

#[derive(Clone, Debug, Default)]
pub struct DidBridgeConfig {
    pub bridge_id: String,
    pub trust_domain: String,
    pub allowed_methods: Option<Vec<String>>,
}

pub struct DidBridge {
    cfg: DidBridgeConfig,
}

impl DidBridge {
    pub fn new(cfg: DidBridgeConfig) -> Self {
        DidBridge { cfg }
    }

    pub fn resolve_did_key(&self, did_url: &str) -> Result<DidDocument, BridgeError> {
        let method = parse_did_method(did_url)?;
        if let Some(allow) = &self.cfg.allowed_methods {
            if !allow.iter().any(|m| m == method) {
                return Err(BridgeError::Rejected(format!(
                    "DID method {} not in allow-list",
                    method
                )));
            }
        }
        if method != "key" {
            return Err(BridgeError::Unsupported(format!(
                "method {} not supported by built-in resolver; provide a custom resolver",
                method
            )));
        }
        let multibase = did_url
            .strip_prefix("did:key:")
            .ok_or_else(|| BridgeError::InvalidInput(format!("not did:key: {}", did_url)))?;
        let id = format!("did:key:{}", multibase);
        Ok(DidDocument {
            id: id.clone(),
            verification_method: Some(vec![DidVerificationMethod {
                id: format!("{}#{}", id, multibase),
                kind: "Ed25519VerificationKey2020".into(),
                controller: id.clone(),
                public_key_multibase: Some(multibase.to_string()),
                public_key_jwk: None,
            }]),
            authentication: Some(vec![Value::String(format!("{}#{}", id, multibase))]),
            controller: Some(Value::String(id)),
        })
    }

    pub fn accept(&self, document: &DidDocument) -> Result<ActorIdentity, BridgeError> {
        let vms = document
            .verification_method
            .as_ref()
            .ok_or_else(|| BridgeError::Rejected("DID has no verification methods".into()))?;
        if vms.is_empty() {
            return Err(BridgeError::Rejected(
                "DID verification_method is empty".into(),
            ));
        }
        let vm = &vms[0];
        let pk = extract_public_key(vm).ok_or_else(|| {
            BridgeError::Unsupported(format!("vm {} has no usable public key", vm.id))
        })?;
        let actor_id = format!(
            "tf:actor:human:{}/{}",
            self.cfg.trust_domain,
            url_encode(&document.id)
        );
        Ok(ActorIdentity {
            identity_version: ActorIdentity_IdentityVersion::V1,
            actor_id,
            actor_type: ActorType::Human,
            instance_id: None,
            public_keys: vec![PublicKey {
                key_id: vm.id.clone(),
                algorithm: pk.algorithm,
                public_key: base64::engine::general_purpose::STANDARD.encode(&pk.bytes),
                purpose: PublicKey_Purpose::Signing,
                valid_from: None,
                valid_until: None,
            }],
            trust_levels: vec![TrustLevel::T2],
            authority_roots: vec![AuthorityRoot {
                kind: AuthorityRoot_Kind::Federation,
                id: vm.controller.clone(),
            }],
            attestations: None,
            valid_from: now_iso8601(),
            valid_until: None,
            revocation_ref: None,
            signature: None,
        })
    }
}

impl Bridge for DidBridge {
    fn bridge_id(&self) -> &str {
        &self.cfg.bridge_id
    }
    fn kind(&self) -> BridgeKind {
        BridgeKind::Did
    }
    fn trust_domain(&self) -> &str {
        &self.cfg.trust_domain
    }
}

struct ProjectedKey {
    algorithm: String,
    bytes: Vec<u8>,
}

fn extract_public_key(vm: &DidVerificationMethod) -> Option<ProjectedKey> {
    if let Some(mb) = &vm.public_key_multibase {
        let decoded = decode_multibase(mb)?;
        if decoded.len() >= 2 && decoded[0] == 0xed && decoded[1] == 0x01 {
            return Some(ProjectedKey {
                algorithm: "ed25519".into(),
                bytes: decoded[2..].to_vec(),
            });
        }
        return Some(ProjectedKey {
            algorithm: vm.kind.to_lowercase(),
            bytes: decoded,
        });
    }
    if let Some(jwk) = &vm.public_key_jwk {
        if jwk.get("kty").and_then(Value::as_str) == Some("OKP")
            && jwk.get("crv").and_then(Value::as_str) == Some("Ed25519")
        {
            if let Some(x) = jwk.get("x").and_then(Value::as_str) {
                let bytes = URL_SAFE_NO_PAD.decode(x).ok()?;
                return Some(ProjectedKey {
                    algorithm: "ed25519".into(),
                    bytes,
                });
            }
        }
    }
    None
}

fn parse_did_method(did: &str) -> Result<&str, BridgeError> {
    let rest = did
        .strip_prefix("did:")
        .ok_or_else(|| BridgeError::InvalidInput(format!("not a DID: {}", did)))?;
    let end = rest
        .find(':')
        .ok_or_else(|| BridgeError::InvalidInput(format!("not a DID: {}", did)))?;
    Ok(&rest[..end])
}

fn url_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char);
            }
            _ => out.push_str(&format!("%{:02X}", b)),
        }
    }
    out
}

fn decode_multibase(s: &str) -> Option<Vec<u8>> {
    if s.is_empty() {
        return None;
    }
    let prefix = s.as_bytes()[0];
    let body = &s[1..];
    match prefix {
        b'z' => base58btc_decode(body),
        b'm' => base64::engine::general_purpose::STANDARD.decode(body).ok(),
        b'u' => URL_SAFE_NO_PAD.decode(body).ok(),
        _ => None,
    }
}

const BASE58_ALPHABET: &[u8] = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

fn base58btc_decode(s: &str) -> Option<Vec<u8>> {
    if s.is_empty() {
        return Some(Vec::new());
    }
    let mut zeros = 0usize;
    while zeros < s.len() && s.as_bytes()[zeros] == b'1' {
        zeros += 1;
    }
    let size = ((s.len() - zeros) as f64 * 0.733).ceil() as usize + 1;
    let mut b256 = vec![0u8; size];
    for i in zeros..s.len() {
        let c = s.as_bytes()[i];
        let idx = BASE58_ALPHABET.iter().position(|&b| b == c)?;
        let mut carry = idx;
        for j in (0..size).rev() {
            carry += b256[j] as usize * 58;
            b256[j] = (carry & 0xff) as u8;
            carry >>= 8;
        }
        if carry != 0 {
            return None;
        }
    }
    let mut start = 0usize;
    while start < size && b256[start] == 0 {
        start += 1;
    }
    let mut out = vec![0u8; zeros];
    out.extend_from_slice(&b256[start..]);
    Some(out)
}

pub fn ed25519_public_key_to_did_key(pub_bytes: &[u8]) -> Result<String, BridgeError> {
    if pub_bytes.len() != 32 {
        return Err(BridgeError::InvalidInput(format!(
            "ed25519 public key must be 32 bytes, got {}",
            pub_bytes.len()
        )));
    }
    let mut prefixed = Vec::with_capacity(2 + 32);
    prefixed.push(0xed);
    prefixed.push(0x01);
    prefixed.extend_from_slice(pub_bytes);
    Ok(format!("z{}", base58btc_encode(&prefixed)))
}

fn base58btc_encode(bytes: &[u8]) -> String {
    if bytes.is_empty() {
        return String::new();
    }
    let mut zeros = 0usize;
    while zeros < bytes.len() && bytes[zeros] == 0 {
        zeros += 1;
    }
    let size = ((bytes.len() as f64) * 1.366).ceil() as usize + 1;
    let mut b58 = vec![0u8; size];
    for &byte in bytes.iter().skip(zeros) {
        let mut carry = byte as usize;
        for j in (0..size).rev() {
            carry += b58[j] as usize * 256;
            b58[j] = (carry % 58) as u8;
            carry /= 58;
        }
    }
    let mut start = 0usize;
    while start < size && b58[start] == 0 {
        start += 1;
    }
    let mut out = String::new();
    for _ in 0..zeros {
        out.push('1');
    }
    for &b in &b58[start..] {
        out.push(BASE58_ALPHABET[b as usize] as char);
    }
    out
}

fn now_iso8601() -> String {
    let secs = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs() as i64;
    let (y, m, d, h, mi, s) = secs_to_ymdhms(secs);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, h, mi, s)
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 {
        (mp + 3) as u32
    } else {
        (mp - 9) as u32
    };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}
